### PR TITLE
v1.3 80 Price histogram

### LINKED ISSUE #80: As a user, I can see a price histogram of the properties and buyers on screen, so I can observe market conditions.
Command: `stat [(property | buyer)]`
Requirements:
- opens new popup with histogram
- `stat (property | buyer)` shows only histogram of property or buyer
- `stat` shows the two histograms overlapped (each column having two columns, one for buyer and one for property)
- must include a legend

### PR DESCRIPTION
**Pull request by trxe**
closes #80.

Summary:
- Added new command `StatCommand` and its parser
- Refactored `CommandResult` to contain an `Optional<UiElement>` field
- `HistogramStat` is a `UiElement` which creates a `JFreeChart` histogram on reaching the `StatWindow` handler.
- `MainWindow` modified as necessary.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/StatCommand.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        requireNonNull(model);
+        ObservableList<Buyer> buyerList = model.getFilteredBuyerList();
+        ObservableList<Property> propertyList = model.getFilteredPropertyList();
+        String msgArgs = ALL_WORD;
+        if (showBuyer && !showProperty) {
+            msgArgs = BUYER_WORD;
+        } else if (!showBuyer & showProperty) {
+            msgArgs = PROPERTY_WORD;
+        }
+        Stat stat = new HistogramStat(buyerList, propertyList, showBuyer, showProperty, msgArgs);
```
Comment (guangxuan) [inline src/main/java/seedu/address/logic/commands/StatCommand.java:50]: we can do some empty list checking here. Throw some exception to be caught by LogicManager.execute

File: src/main/java/seedu/address/logic/parser/StatCommandParser.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    @Override
+    public StatCommand parse(String userInput) throws ParseException {
+        requireNonNull(userInput);
+        Actor actor;
+        userInput = userInput.trim();
+        try {
+            ParserUtil.assertPreambleArgsCount(userInput, NUM_OF_PREAMBLE_ARGS);
+            actor = ParserUtil.parseActor(userInput, ACTOR_POSITIONAL_INDEX);
+        } catch (ParseException pe) {
+            if (userInput.equals(new String())) {
```
Comment (guangxuan) [inline src/main/java/seedu/address/logic/parser/StatCommandParser.java:27]: Slight nitpicking here but we could use .isEmpty()

File: src/main/java/seedu/address/model/stats/HistogramStat.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public JFreeChart create() {
+        double[] buyerPrices = buyerList.stream()
+                .mapToDouble(buyer -> (double) buyer.getMaxPrice().value).toArray();
+        double[] propertyPrices = propertyList.stream()
+                .mapToDouble(property -> (double) property.getPrice().value).toArray();
+
+        HistogramDataset dataset = new HistogramDataset();
+        if (showBuyer) {
+            dataset.addSeries("buyers", buyerPrices, BIN_COUNT);
```
Comment (guangxuan) [inline src/main/java/seedu/address/model/stats/HistogramStat.java:48]: empty lists seem to cause exceptions here. 
Interestingly `stat` is not NULL within MainWindow but it just cannot draw here.
Exception in thread "JavaFX Application Thread" java.lang.IllegalArgumentException: Null or zero length 'values' argument.
	at org.jfree.data.statistics.HistogramDataset.getMinimum(HistogramDataset.java:198)
	at org.jfree.data.statistics.HistogramDataset.addSeries(HistogramDataset.java:113)
	at seedu.address.model.stats.HistogramStat.create(HistogramStat.java:48)
	at seedu.address.ui.StatWindow.show(StatWindow.java:60)
	at seedu.address.ui.MainWindow.handleStat(MainWindow.java:173)

File: src/main/java/seedu/address/ui/UiElement.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package seedu.address.ui;
+
```
Comment (guangxuan) [inline src/main/java/seedu/address/ui/UiElement.java:2]: Could add a javadoc here. Short one will do.

File: src/main/java/seedu/address/model/stats/Stat.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package seedu.address.model.stats;
+
+import org.jfree.chart.JFreeChart;
+
+import seedu.address.ui.UiElement;
+
```
Comment (guangxuan) [inline src/main/java/seedu/address/model/stats/Stat.java:6]: Could add a javadoc here. Short one will do.

File: src/main/java/seedu/address/model/stats/HistogramStat.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package seedu.address.model.stats;
+
+import org.jfree.chart.ChartFactory;
+import org.jfree.chart.JFreeChart;
+import org.jfree.data.statistics.HistogramDataset;
+
+import javafx.collections.ObservableList;
+import seedu.address.model.property.Buyer;
+import seedu.address.model.property.Property;
+
```
Comment (guangxuan) [inline src/main/java/seedu/address/model/stats/HistogramStat.java:10]: Could add a javadoc here.

File: src/main/java/seedu/address/model/stats/HistogramStat.java (hunk 3 above)
Comment (trxe) [reply src/main/java/seedu/address/model/stats/HistogramStat.java:48]: Alright sure, handling empty lists as explained in the comment above.

File: src/main/java/seedu/address/logic/commands/StatCommand.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
+import javafx.collections.ObservableList;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+import seedu.address.model.property.Buyer;
+import seedu.address.model.property.Property;
+import seedu.address.model.stats.HistogramStat;
+import seedu.address.model.stats.Stat;
+
```
Comment (guangxuan) [inline src/main/java/seedu/address/logic/commands/StatCommand.java:13]: oh I forgot about this javadocs too

**Comment by trxe** [comment]
New updates:
- Added Priceable interface to Buyer and Property for easier handling in statistics.
- Fixed visual bugs for histogram by implementing creation and labelling of bins.

File: src/main/java/seedu/address/model/stats/HistogramStat.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    /**
+     * Creates a histogram to displayed as a popup.
+     * Shows either buyers, properties or both.
+     *
+     * @return price histogram.
+     */
+    public JFreeChart create() {
+        assert !buyerList.isEmpty() && !propertyList.isEmpty();
```
Comment (guangxuan) [inline src/main/java/seedu/address/model/stats/HistogramStat.java:107]: btw this assertion fails when we try to do `stat buyer` with no `properties`
or `stat` with no `properties`

**Review by guangxuan — CHANGES_REQUESTED** [review]
pretty close already, this should be the last bug

**Pull request merged**